### ISSUE TITLE
preferred( ) yields unexpected results

### ISSUE THREAD

**Issue by gwselke** [manwar/Map-Tube#14]
While the documentation for what the method preferred( ) should be doing is very brief, I assume it is supposed to remove the display of lines that exist at certain stations but that cannot be meaningfully used within a given route. (If this understanding is not correct, please, let me know!)
However, while preferred( ) does remove some lines from some stations, it also displays lines for stations which are never actually touched by that line.

Example:

`map-tube --map=KoelnBonn --start="Dom / Hbf" --end="Tannenbusch Mitte"`

correctly shows a route starting with "Dom / Hbf (16, 18, 5, hbfkoeln)" and ending in "Tannenbusch Mitte (16, 63)". Indeed, line 16 provides the shortest path from start to end, and the other lines mentioned do pass their respective stations.
However,

`map-tube --map=KoelnBonn --start="Dom / Hbf" --end="Tannenbusch Mitte"  --preferred`

starts with "Dom / Hbf (16, 18)" and ends in "Tannenbusch Mitte (16, 17)". For one, it is not clear why the starting point displays "18". More importantly, line 17 (displayed for the target station) actually does not run there at all. (Compare with the "non-preferred" version above, which correctly shows 16 and 63 but not 17.)

**Comment by manwar** [comment]
@gwselke this is great, so the issue is with how `--preferred` is working now? I have another question, does this behaviour repeatable with any other map data e.g. London? If yes then definitely we need to look at core code with regard to how we deal with it. Even bigger question, does this issue cropped up recently or it was there long time ago and we only noticed now.

**Comment by gwselke** [comment]
As for "when did it start": Good question :-) I just recently came back to the maps, so I cannot say when it started. Also, I am not sure whether I fully understand what --preferred is supposed to do. Could you, please, describe in more detail? I would certainly be willing to look at the code, but first I think I should understand better (rather than assume that my expectatiosn are identical to yours).

As for "does it occur with other maps": Good idea, I'll try to reproduce for London. Either I can (then it would be an issue with Map::Tube), or I can't (in which case I'd have to have a serious talk to the author of Map::Tube::KoelnBonn, i.e., to myself). Gimme a little time.

**Comment by gwselke** [comment]
I can now confirm that it is a (small, but consequential) bug in Map::Tube (Map::Tube::Utils, to be exact), and that it also shows in Map::Tube::London. It only shows in routes that contain more than 10 stations. 

The problem is in method `filter( )` defined in Map::Tube::Utils. This method takes, as input, a list of lists. Each entry in the list corresponds to one station on the route (in traversal order), and each such entry is the list of lines that touches on that station.
The method should return exactly such a list of lists (in the same order!), but each entry will contain only lines that are also serving the previous or the next station on this route. (Thus, only lines that could be used on that part of the route.)
The method builds an intermediate hash `%r` which contains as keys the index of the station and as value the "useful" lines at that station. This index is *not* the station id but rather the position of that station on the route, as present in the input argument, i.e., the keys of `%r` are integer values.
The return value is calculated by this statement:

```
   return [
        map [ sort { $a cmp $b }
              keys %{ $r{$_} } ],
        sort { $a cmp $b } keys %r
    ];    
```
The bug is in the part `sort { $a cmp $b } keys %r`, which erroneously sorts the (integer) keys alphabetically -- which messes up the proper order: when there are, say, 14 entries( (0..13), the resulting order will be 0, 1, 10, 11, 12, 13, 2, 3, ...
The quick and easy fix is to use `<=>` instead of `cmp` in order to make a numeric comparison in sorting.
(Note that the other sort is operating on line ids (i.e., strings), so that one is correctly doing an alphabetical sort.) 
So the resultant code would be 

```
   return [
        map [ sort keys %{ $r{$_} } ],
        sort { $a <=> $b } keys %r
    ];    
```
(Note that the `{ $a cmp $b }` on the first sort can be omitted, because this is the default comparison used by sort.)

A different approach would be to replace the hash `%r` with an array, which would obviate the need for sorting these keys.
(The other sort would still be needed, though.)

I am uploading a test script (renamed to file extension .txt due to github restrictions on uploads) which exhibits the problem in the current version of Map::Tube and which should pass after fixing. If you want, feel free to include this in the test suite for Map::Tube (or Map::Tube::London --but I think it belongs into the former, since the problem is not with the London data.)
[preferred-bug.t.txt](https://github.com/user-attachments/files/17963375/preferred-bug.t.txt)

**Comment by manwar** [comment]
@gwselke This is brilliant, you nailed it. Thanks for your contribution as always. I will do quick review and merge your changes soon.

**Comment by manwar** [comment]
@gwselke Just to inform you, I have released your patch as v3.78.
https://metacpan.org/release/MANWAR/Map-Tube-3.78

**Comment by gwselke** [comment]
Thank you, @manwar! That was extremely fast action!
